[JENKINS-75675] Refactor class loading logic in order to reduce memory consumption

dukhlov opened the pull request:
See [JENKINS-75675](https://issues.jenkins.io/browse/JENKINS-75675).

### Background
Our Jenkins installation relies heavily on Groovy pipelines, shared libraries, and similar mechanisms. Groovy uses dynamic invocation via MetaClass, CallSite, etc. During introspection, it attempts to load many classes that don’t exist, simply to determine whether a given token is a class, a variable, or something else.
My investigation shows that this behavior leads to memory leakage.

### Problem
The base ClassLoader implementation in Java (which all other class loaders inherit from) supports two modes:

parallelCapable (currently used by all Jenkins core class loaders, including WebAppClassLoader and PlatformClassLoader)
non-parallel (legacy mode)
Parallel-capable class loaders create and retain a lock object per class name, indefinitely. So, if we have 1,000 class loading misses, every parallel-capable class loader in the hierarchy keeps 1,000 unused lock objects forever.

In the case of Jenkins's UberClassLoader, a typical setup with ~200 plugins results in around 500 class loaders. In our Jenkins instance, about 2,000 classes are loaded successfully, while we observe over 200,000 class loading misses.

As a result, the class loaders retain:

500 class loaders × 200,000 lock objects each
Plus the internal ConcurrentHashMap.Node objects used to store them
This results in roughly 2 GB of unnecessary memory consumption.

### Solution
- Introduce base `DelegatingClassLoader` implementation which overrides base `ClassLoaders.loadClass` to remove not needed base locking mechanism (it don't load class itself just delegates loading to another class loader and locking will be done here)
- extend `DelegatingClassLoader` instead of `ClassLoader` if possible.
- override `getClassLoadingLock()` for `UrlClassLoader2` for use GCed locking object which won't be kept forever
- Introduce FilteringClassLoader which can check if class exists (using `getResource()` call) and don't use `loadClass()` call not to create locking Object at all
- wrap base `WebAppClassLoader` with FilteringClassLoader to avoid not needed lock files creation

### Testing done

This PR tested by running 2-hour long proprietary automation tests based on jenkins scripted pipeline jobs
class loading logic is broadly used, so it definitely covered by this testing

Also I made `jcmd GC.class_histogram`. Amount of `java.lang.Object` and `java.util.concurrent.ConcurrentHashMap$Node` instances was decreased dramatically

Before fix:
<img width="769" alt="Screenshot 2025-05-19 at 06 26 27" src="https://github.com/user-attachments/assets/852fa965-0dbd-4df4-9059-6e895063e919" />

After fix:
<img width="906" alt="Screenshot 2025-05-26 at 19 59 19" src="https://github.com/user-attachments/assets/20a100ca-a481-4fee-a8ef-8f5b178eb3bd" />


### Proposed changelog entries

- Reduce Plugin's ClassLoader memory consumption.

### Proposed changelog category

/label internal

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] The Jira issue, if it exists, is well-described.
- [x] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)). Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [x] There is automated testing or an explanation as to why this change has no tests.
- [x] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [x] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [x] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [x] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [x] For new APIs and extension points, there is a link to at least one consumer.

### Desired reviewers

@jenkinsci/core-pr-reviewers

Before the changes are marked as `ready-for-merge`:

### Maintainer checklist

- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).

basil commented:
Would this not be better reported/patched against OpenJDK's `java.net.URLClassLoader`? While we could work around the issue as in this PR, I'm not sure we want to maintain custom class loading logic like this in the long term. We previously spent a long time maintaining a fork of Ant's class loader, which had various issues, so I would definitely prefer to use unmodified upstream OpenJDK code if possible.

dukhlov commented:
Alternative: It is possible to make jenkins.util.URLClassLoader2  not capable for parallel class loading by removing 
```
static {
   registerAsParallelCapable();
}
```

In this case loading classes will be serial, URLClassloader will use sing lock instead of dedicated per class name lock.
But I am not aware of whole set of usecases and don't know do we need parallel class loading in scope of single plugin or not

basil commented:
> This isn’t an issue with URLClassLoader itself — it’s designed to work correctly in typical Java applications, where an extra 10MB per class loader is a reasonable overhead in exchange for a simpler and faster implementation. The problem becomes significant in pluggable systems that use separate class loading.

I don't see anything in the `URLClassLoader` Javadoc that would confirm this viewpoint. The opposite assertion (that it ought to be efficient in pluggable systems) could be made without contradicting the Javadoc.

I hate wasting memory and I am not against this PR as a short-term solution, but only after we have a discussion with the OpenJDK folks about whether or not this is an upstream problem and, if so, what the long-term plan would be.

dukhlov commented:
> I don't see anything in the URLClassLoader Javadoc that would confirm this viewpoint. The opposite assertion (that it ought to be efficient in pluggable systems) could be made without contradicting the Javadoc.
Right, this is just my understanding.  I researched existing standard ClassLoader implementation and don't have arguments to ask OpenJdk team to improve standard implementation which satisfies needs of 99% of java applications just for Jenkins and perhaps about 10 another pluggable java based platforms.
But talking to OpenJdk folks definitely make sense if it is possible.

basil reviewed (dismissed):
https://github.com/jenkinsci/jenkins/pull/10659#issuecomment-2891902245

jtnord commented on core/src/main/java/jenkins/util/URLClassLoader2.java:
```diff
@@ -71,6 +77,6 @@ public Class<?> findLoadedClass2(String name) {
 
     @Override
     public Object getClassLoadingLock(String className) {
-        return super.getClassLoadingLock(className);
+        return parallelLockWeakValuesMap.computeIfAbsent(className, k -> new Object());
```
I wonder why the JDK does not return an `intern`ed String and needs to keep the lock in a map.
For sure interning will put everything in the string pool, but if there are no references then the a String in the string pool should be eligible to be GC'ed  (normally a Sting literal in a class would cause a reference for as long as the class was loaded, meaning it would never actually be GCed, but for explicitly interned Strings this is different)

e.g

```java
  return className.intern();
```

There may be some valid reasons here - interning is not free and could cause a slowdown on classloading.

dukhlov replied:
className.intern() is jvm unique for class name. they need per classloader uniqueness

dukhlov replied:
Javadoc:
* <p> Class loaders that support concurrent loading of classes are known as
 * <em>{@linkplain #isRegisteredAsParallelCapable() parallel capable}</em> class
 * loaders and are required to register themselves at their class initialization
 * time by invoking the {@link
 * #registerAsParallelCapable ClassLoader.registerAsParallelCapable}
 * method. Note that the {@code ClassLoader} class is registered as parallel
 * capable by default. However, its subclasses still need to register themselves
 * if they are parallel capable.
 * In environments in which the delegation model is not strictly
 * hierarchical, class loaders need to be parallel capable, otherwise class
 * loading can lead to deadlocks because the loader lock is held for the
 * duration of the class loading process (see {@link #loadClass
 * loadClass} methods).

dukhlov replied:
They need per class loader + class name unique lock to avoid possible deadlocks.

jtnord replied:
`(System.identityHashCode(this) + "-" + className).intern()`

`-` is an illegal character in the className, and the identity hashcode of the classloader should be unique for as long as it matters (if it is re-used it is because the older classloader has been GC'ed and is no longer in existence so can not be classloading, so can not have a lock - that is unless your JVM somehow crashed during classloading and did not release locks, but at that point you probably have bigger issues)

Now the performance aspects of this may be somewhat questionable. StackOverflows have a habit of not releasing locks so this is probably a hack to far, and for an interim solution the guava map would seem to be much safer.

dukhlov replied:
Does Jenkins use circular pependent clasloders? If not the easiest solution is disabling parallel classloading for plugin's classloaders. I tried on my test cluster and not noticed any changes in speed. The code I tried:
```

/**
 * {@link URLClassLoader} with loosened visibility for use with {@link
 * ClassLoaderReflectionToolkit}.
 */
@Restricted(NoExternalUse.class)
public class URLClassLoader2 extends URLClassLoader implements JenkinsClassLoader {
    public static final boolean DISABLE_PARALLEL_LOADING = Boolean.getBoolean(
            System.getProperty(URLClassLoader2.class.getName() + ".DISABLE_PARALLEL_LOADING"));

    static {
        if (!DISABLE_PARALLEL_LOADING) {
            registerAsParallelCapable();
        }
    }

```

basil replied:
> Also I sent a mail to `core-libs-dev@openjdk.org`, but no answer yet

In the past, we've had success reporting issues to the public [JDK Bug Database](https://bugs.openjdk.org/) using [this form](https://bugreport.java.com/bugreport/).

jtnord replied:
> Does Jenkins use circular pependent clasloders?

pependent?

it doesn't use circular classloaders, any circles are supposed to be broken https://github.com/jenkinsci/jenkins/blob/7aa9647a35bc382c7f632bfac8af5a7d9cd5e6f7/core/src/main/java/hudson/PluginManager.java#L507-L534 but it does use heiracial ones where 2 different plugins would be attempting to load different classes from a third plugin at the same time.

jtnord replied:
> Also I sent a mail to core-libs-dev@openjdk.org, but no answer yet

https://mail.openjdk.org/pipermail/core-libs-dev/2025-May/146225.html

dukhlov replied:
> - is an illegal character in the className, and the identity hashcode of the classloader should be unique for as long as it matters (if it is re-used it is because the older classloader has been GC'ed and is no longer in existence so can not be classloading, so can not have a lock - that is unless your JVM somehow crashed during classloading and did not release locks, but at that point you probably have bigger issues)

Clear now. Yes it will work but in fact give us nothing becaue we stiil need to keep the same amount String objects in string pool, just in one map instead of per class loader map, and even worse, Object instance eats just 16 bytes, String  with class name and hash is several times larger

jtnord replied:
> Yes it will work but in fact give us nothing

not so as the String Pool can be GCed.  so when nothing is referring to the lock it can be removed.

So whilst the class is loading there will be a reference (as it is the lock).  Afterwards there is no reference (or at least once all the attempts to load the same class have finished) , so the string should be eligible for reclamation by the GC from the string pool.

>  just in one map instead of per class loader map, 

there is no map...  there is no lookup...

dukhlov replied:
> but it does use hierarchical ones where 2 different plugins would be attempting to load different classes from a third plugin at the same time.

In this case it will make jenkins initial load slower. At least we shoudn't get deadlock. Perhaps it is possible to make it configurable at least

dukhlov replied:
> there is no map... there is no lookup...

yes, now understand what you mean. Didn't know that stringPool elements can be GCed

timja replied:
https://mail.openjdk.org/pipermail/core-libs-dev/2025-May/146392.html
> This is an issue for "parallel capable" class loaders, esp. those that 
don't directly delegate. I don't know what topology is used in the 
Jenkins plugin class loaders and whether these class loader need to 
parallel capable.
>
> Expunging entries isn't really feasible, at least not without changing 
the spec.
>
> [JDK-8005233](https://bugs.openjdk.org/browse/JDK-8005233) has a write-up from David Holmes on this topic and a 
prototype for "fully concurrent" class loaders. This work has been 
parked for several years. Maybe some day it will get dusted off and a 
JEP drafted (due to the significance).

Anyone know if we need the parallel capable class loader? I have vague memories of Basil fixing it up when unforking the ant class loader we used to use so that we could use the parallel version.

dukhlov replied:
> Hi Dmytro,
> 
> answering offlist.
> 
> That is fascinating to learn that parallelLockMap never shrinks, and it
> also will keep the lock objects for classes loaded by parent loaders and
> even when not being able to find the class.
> > It seems that we need a way to
> > determine whether a class can be loaded from the parent class loader
> > chain
> You should be able to find out by calling
> parentClassLoader.getResource(className.replace('.','/') + ".class") to
> do exactly that - either in the Jenkins or probably better in the Groovy
> classloader.
> Would this work? Of course this adds overhead...

dukhlov replied:
How tomcat solves similar problem:
https://github.com/apache/tomcat/blob/9432ba70546820dcda27738c932c828282cd676a/java/org/apache/catalina/loader/WebappClassLoaderBase.java#L1098

basil replied:
Yes, [JDK-8005233](https://bugs.openjdk.org/browse/JDK-8005233) covers this problem well. Looks like the problem is well understood by the OpenJDK developers, and fixing this properly upstream would require a Java Enhancement Proposal (JEP) and change to the specification.

> groovy.tmp.templates.GStringTemplateScript105342BeanInfo:Optional.empty, groovy.tmp.templates.GStringTemplateScript125196Customizer:Optional.empty, java.lang.hudson.model.params$SLACK_ALERT_WEBHOOK:Optional.empty, groovy.tmp.templates.GStringTemplateScript6444BeanInfo:Optional.empty, groovy.tmp.templates.GStringTemplateScript15126Customizer:Optional.empty, groovy.tmp.templates.GStringTemplateScript41396BeanInfo:Optional.empty, java.util.jenkins.model.params$IS_INTERNAL_REGISTER:Optional.empty, groovy.tmp.templates.GStringTemplateScript23236BeanInfo:Optional.empty,  
after fix - this numper is ~2k and not growing

https://github.com/jenkinsci/jenkins/blob/608d83cc4b2cab4d108500d41c509d4ae5f71550/core/src/main/java/hudson/PluginManager.java#L2416-L2418 already contains exclusions for Groovy garbage. I wonder if we need to expand what is in that list to include the classes you discovered?

basil replied:
> Anyone know if we need the parallel capable class loader?

If I remember correctly, enabling parallel loading on at least some of the Jenkins class loaders resulted in a significant performance improvement. So I went ahead and enabled it everywhere. If disabling it on some of the classloaders results in a big improvement in memory usage without a significant performance regression, I would be OK with that.

jtnord reviewed (commented):
Sorry forgot to submit this earlier.

I Agree with Basil, this appears to be something for the JDK. 
Anything here should be a temporary workaround until we could pickup a JDK with a fix (should the JDK team agree this is an issue)

dukhlov commented on core/src/main/java/jenkins/util/URLClassLoader2.java:
```diff
@@ -1,7 +1,9 @@
 package jenkins.util;
 
+import com.google.common.collect.MapMaker;
```
Understood

Commit:
JENKINS-75675: Refactor classloader logic in order to reduce memory consumption

Override URLClassloader2.getClassLoadingLock to use lock objects that can be GCed when not needed,
Wrap WebAppClassLoader with CheckingExistenceClassLoader to avoid missing loadClass() calls not to crate unnecessary persistent lock object,
Refactor delegating classloaders not to use locking:
 * DelegatingClassloader - non-locking base implementation and use it for classloaders
   which delegate functionality to other classloaders,
 * CheckingExistenceClassLoader - check that class exists before loadClass() calls.

dukhlov commented:
> Can you update the description please with re-running the testing that you did before?

Done, re-run my tests and updated 'after fix' jcmd screen shot

basil commented:
Thanks, this is excellent! Can you please write a description of the changes to help with reviewing?

basil commented:
Can we keep as many of the calls to `registerAsParallelCapable` as possible while still achieving the goal of this PR, which is to decrease memory usage?

dukhlov commented:
> Can we keep as many of the calls to `registerAsParallelCapable` as possible while still achieving the goal of this PR, which is to decrease memory usage?

It is make sense to cal registerAsParallelCapable for URLClassLoader2 and PluginFirstClassLoader2 only, the others don't use locking

What I see on my env - typical plugins loads ~1k (previously was 100k * 2 (because DepencyClasLoader stores its own parallelLockMap)) classes, it will add ~40kb (previously 8mb) memory overhead for parallelLockMap, if we have 200 plugins - ~8mb overhead (previously was 1,6gb) - looks like something that we can afford

But... Jenkins Apllication classloader is not parallel capable https://github.com/dekellum/jetty/blob/master/jetty-webapp/src/main/java/org/eclipse/jetty/webapp/WebAppClassLoader.java#L382
and the main set of class loading will be locked on it. Therefore don't know how noticeable will be effect of calling registerAsParallelCapable on URLClassLoader2 and PluginFirstClassLoader2.

Also want to say that If some plugin is written using groovy, not java it will be a problem... This patch fix problem for uberClassLoader only. Implementing such approach for each plugin's and using cache for loaded class and  class loading misses is expensive  because of scaling with plugin number.  Anyway it is another story.

Should I revert registerAsParallelCapable for URLClassLoader2 and PluginFirstClassLoader2?

basil commented:
Thanks. I appreciate the description of the implementation-level changes to each file. I think what I'm still missing is a high-level description of the overall change, and how that solves the problem of excess memory consumption.

dukhlov commented:
> Thanks. I appreciate the description of the implementation-level changes to each file. I think what I'm still missing is a high-level description of the overall change, and how that solves the problem of excess memory consumption.

We have an UberClassLoader. When we call loadClass on it, it's possible that loadClass will be invoked on each underlying classloader (e.g., WebAppClassLoader, all DependencyClassLoaders, and all URLClassLoader2 instances) — in the worst case, if the class does not exist.

For each loadClass call on the UberClassLoader, up to 2 * plugin_count + number_of_classloaders_in_the_parent_chain lock objects may be created, which won’t be garbage collected.

The problem I’m facing is that Groovy pipelines generate a large number of loadClass calls for non-existent classes (some class names can be found in the comments above — some of them are dynamically generated).
As a result, the parallelLockMap size for each classloader can grow to 100,000 entries.

The initial solution was to use weak references for lock objects, which helps mitigate the issue.
(Disadvantage: the WebAppClassLoader chain still uses strong references, but this approach prevents the issue from scaling with the number of plugins.)

Then we found a better solution — call getResource(className + ".class") before loadClass, to avoid calling loadClass for non-existent classes. This prevents the creation of persistent lock objects for classes that don’t exist.

In summary:
The main change is to filter out non-existent classes before calling uberClassLoader.loadClass().

Additional improvement 1: introduce DelegationClassLoader to completely eliminate unnecessary locking.

dukhlov commented:
This solution in my opinion is reasonable compromise. 

What can be improved next:

1. Override getClassLoadingLock() in URLClassLoader2 to return weak references. This would further reduce memory overhead by allowing unused lock objects to be garbage collected.
2. Apply getResource(className + ".class") filtering independently for each underlying classloader. Currently, for example, if a class exists in a plugin’s classloader, we still attempt to load it from the WebAppClassLoader, resulting in an unnecessary class loading miss, WebAppClassloader will have lockObjects for each classes available in whole UberClassLoader

dukhlov commented:
@basil, I developed such overriding for getClassLoadingLock which does not use guava, but didn't push it because as far I understood you don't want to maintain such code

```
@Restricted(NoExternalUse.class)
public class URLClassLoader2 extends URLClassLoader implements JenkinsClassLoader {
    private final WeakHashMap<String, WeakReference<String>> parallelLockMap2 = new WeakHashMap<>();
 
    @Override
    @SuppressFBWarnings(value = "DM_STRING_CTOR", justification = "Need to have a lock object which can be GCed and " +
            "can be retrieved by className. Solution: create new String(className) object, use it as lockObject " +
            "and store in WeakHashMap as a key and weak referenced value. It can be retrieved by className because " +
            "className.hashcode() == new String(className).hashcode(), className.equals(new String(className)) == true")
    public Object getClassLoadingLock(String className) {
        synchronized (parallelLockMap2) {
            WeakReference<String> lockObjectWeakRef = parallelLockMap2.get(className);
            String lockObject = null;
            if (lockObjectWeakRef != null) {
                lockObject =  lockObjectWeakRef.get();
            }
            if (lockObject != null) {
                return lockObject;
            }
            lockObject = new String(className);
            parallelLockMap2.put(lockObject, new WeakReference<>(lockObject));
            return lockObject;
        }
    }
}
```

dukhlov commented:
@basil, @timja , @jtnord, @qerub, I would like to understand whether this patch aligns with the core developers' vision for the direction of the project. Should I change something, or would it be better to stop and close the PR if I'm heading in the wrong direction?

timja commented:
> @basil, @timja , @jtnord, @qerub, I would like to understand whether this patch aligns with the core developers' vision for the direction of the project. Should I change something, or would it be better to stop and close the PR if I'm heading in the wrong direction?

It definitely looks useful and it does look good to me.

basil commented:
@dukhlov I have been travelling this week and have not had enough time to work on this PR. I want to get this shipped, but this PR is still a bit too big and difficult for me to understand. Even after asking for explanations, it seems that this PR trying to solve several problems at once. I would prefer if it were split into two PRs, one being the minimal changes needed to directly address the memory usage issue, and another PR with any other cleanups. Describing the change(s) with the template [here](https://raw.githubusercontent.com/delphix/.github/refs/heads/master/pull_request_template.md) would also help me process this more quickly. If these things aren't done, that isn't necessarily a blocker for this work. I will still get to it eventually, but the lack of the above would place more work on me and it would take longer.

dukhlov commented:
> I am working on some performance improvements for job loading, but this is next on my TODO list after that. Hopefully by next week.

Ah, great, thank you!

basil committed:
Fast incremental

timja commented:
(@basil friendly reminder if you are able to review this it would be appreciated.)

dukhlov commented on core/src/main/java/hudson/PluginManager.java:
```diff
@@ -2392,25 +2394,28 @@ public <T> T of(String key, Class<T> type, Supplier<T> func) {
-            if (name.startsWith("SimpleTemplateScript")) { // cf. groovy.text.SimpleTemplateEngine
-                throw new ClassNotFoundException("ignoring " + name);
+            String[] namePrefixesToSkip = new String[] {
+                    "groovy.tmp.",          // temporary groovy classes that expected in scope of GroovyClassLoader only
```
This change was reverted since it’s part of a different PR’s scope.

basil committed:
Merge branch 'master' into feature/JENKINS-75675

basil committed:
grammar: Rename `ExistenceCheckingClassLoader` to `CheckingExistenceClassLoader`

basil committed:
Mark as restricted

basil committed:
Additional null checks

basil committed:
Mark as final

basil committed:
Add verification

basil committed:
Clean up `ExistenceCheckingClassLoader`

basil committed:
Clean up `DelegatingClassLoader`

basil committed:
Improve documentation

basil committed:
Wrap lines at reasonable length

basil committed:
Make `DelegatingClassLoader` abstract

basil reviewed (commented):
Thanks very much! Sorry for the delay in reviewing this. I just took a look and added some commits with small improvements. Feel free to revert any of them if you think it makes your PR worse.

GPT-5 suggested using a per-instance monotonically increasing ID in `getClassLoadingLock` to avoid `identityHashCode` collision, as well as precomputing a per-instance prefix once to avoid rebuilding it each time. I'll leave it up to you as to whether you think one or both of those suggestions is valid.

Can you please test the `ExistenceCheckingClassLoader` portion with both Tomcat and Winstone? We only officially support Winstone, but a small percentage of Jenkins users still use Tomcat, and we try to have no regressions there. If this is asking too much, I can look into this myself.

I will also run a wider set of automated tests on this PR and report the results here.

Commit:
Improve lockObject string calculation

basil committed:
Revert "Fast incremental"

This reverts commit 6ad9f3b665a9a53a3c498f50664d573c7154a773.

basil reviewed (approved):
Very nice! I tested this interactively and verified the memory savings as in the **Testing Done** section. I also ran PCT and ATH successfully. The former isn't a very realistic test since it does not use these class loaders, and the latter is realistic but plugin coverage is low. The author of the PR says it has been deployed in production successfully, so that gives me confidence these changes are correct. I see no regressions, either functional or performance-wise.

timja reviewed (approved):
This looks like a great solution, thanks for bearing with us throughout the process.

Thanks Basil for doing the additional testing and reproducing the results

---

/label ready-for-merge

----

This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

A1exKH reviewed (approved):
LGTM.

basil commented:
(I plan to squash-merge this with @jtnord listed as a co-author as well, because he came up with the important idea of the string interning lock.)

Commit:
Add mail.openjdk.org link

dukhlov committed:
Merge branch 'master' into feature/JENKINS-75675

Commit:
Add mail.openjdk.org link. Fix.

basil commented:
I ran the Tomcat tests from https://github.com/jenkinsci/packaging/tree/master/molecule/servlet with the WAR from this PR, and the test passed, so I think this PR is ready for merge.

basil committed:
Simplify method

basil committed:
Clean up comment

basil committed:
Clean up formatting
